File: TMOAncuti19/train_ICA.py
import cv2
import numpy as np
from sklearn.decomposition import FastICA
import os
import glob
import yaml

patch_size = 31
num_components = 10
patches_per_image = 100
dataset_dir = '/TMS_build/images_subset'

# Save model in YAML OpenCV format.
def save_model_opencv_format(filename, icaBasis, mu, sigma):
    with open(filename, "w") as f:
        f.write("%YAML:1.0\n")
        f.write("---\n")
        
        def write_matrix(key, mat):
            rows, cols = mat.shape
            dt = "d"
            data_str = ", ".join(map(str, mat.flatten()))
            f.write(f"{key}: !!opencv-matrix\n")
            f.write(f"   rows: {rows}\n")
            f.write(f"   cols: {cols}\n")
            f.write(f"   dt: {dt}\n")
            f.write(f"   data: [ {data_str} ]\n")
        
        write_matrix("icaBasis", icaBasis)
        write_matrix("mu", mu)
        write_matrix("sigma", sigma)

# Extract patches from an image.
def extract_patches(image, patch_size, num_patches):
    patches = []
    # Convert to grayscale if needed.
    if len(image.shape) == 3:
        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    else:
        gray = image.copy()
    # Normalize to [0,1] and convert to float64.
    gray = gray.astype(np.float64) / 255.0
    h, w = gray.shape
    for _ in range(num_patches):
        r = np.random.randint(0, h - patch_size)
        c = np.random.randint(0, w - patch_size)
        patch = gray[r:r+patch_size, c:c+patch_size]
        patches.append(patch.flatten())
    return np.array(patches)

# Collect patches from a dataset.
all_patches = []
image_files = glob.glob(os.path.join(dataset_dir, '*.*'))
for filepath in image_files:
    image = cv2.imread(filepath)
    if image is None:
        continue
    patches = extract_patches(image, patch_size, patches_per_image)
    all_patches.append(patches)
all_patches = np.vstack(all_patches)
print("Total patches extracted:", all_patches.shape[0])

# Run FastICA.
ica = FastICA(n_components=num_components, random_state=0)
ica.fit(all_patches)
ica_basis = ica.components_

# Compute the independent coefficients for all patches.
coeffs = ica.transform(all_patches)
# Compute mu and sigma from the independent coefficients.
mu = np.mean(coeffs, axis=0).reshape(-1, 1)
sigma = np.std(coeffs, axis=0).reshape(-1, 1)

# Save the model
save_model_opencv_format("ica_model.yml", ica_basis, mu, sigma)
print("ICA model saved to ica_model.yml")

File: TMOAncuti19/train_ICA/train_ICA.cpp
#include <opencv2/opencv.hpp>
#include <itpp/itbase.h>
#include <itpp/signal/fastica.h>
#include <itpp/stat/misc_stat.h>
#include <iostream>
#include <filesystem>
#include <fstream>
#include <random>

namespace fs = std::filesystem;
using namespace itpp;
using namespace std;

static constexpr int PATCH_SIZE = 31;
static constexpr int NUM_COMPONENTS = 10;
static constexpr int PATCHES_PER_IMG = 100;
const string DATASET_DIR = "/TMS_ica_images";
const string OUTPUT_FILE = "../ica_model.yml";

// Extract patches from an image
mat extract_patches(const cv::Mat& gray, int patch_size, int count) {
    int rows = gray.rows - patch_size;
    int cols = gray.cols - patch_size;
    mat patches(count, patch_size * patch_size);

    mt19937 rng{ random_device{}() };
    uniform_int_distribution<> dr(0, rows), dc(0, cols);

    for (int i = 0; i < count; ++i) {
        int r = dr(rng), c = dc(rng);
        cv::Mat block = gray(cv::Rect(c, r, patch_size, patch_size));
        block.convertTo(block, CV_64F, 1.0 / 255.0);
        vec v(patch_size * patch_size);
        memcpy(v._data(), block.ptr<double>(), v.size() * sizeof(double));
        patches.set_row(i, v);
    }
    return patches;
}

// Save model in YAML OpenCV format
void save_yaml(const string& fname, const mat& basis, const vec& mu, const vec& sigma) {
    ofstream f{ fname };
    auto write_mat = [&](const string& key, const mat& m) {
        f << key << ": !!opencv-matrix\n"
          << "   rows: " << m.rows() << "\n"
          << "   cols: " << m.cols() << "\n"
          << "   dt: d\n"
          << "   data: [ ";
        for (int i = 0; i < m.rows(); ++i)
            for (int j = 0; j < m.cols(); ++j)
                f << m(i,j)
                  << ((i==m.rows()-1 && j==m.cols()-1) ? " ]\n" : ", ");
    };
    f << "%YAML:1.0\n---\n";
    write_mat("icaBasis", basis);

    // Write mu and sigma
    mat mu_m   = reshape(mu,   mu.size(),    1);
    mat sig_m  = reshape(sigma, sigma.size(), 1);
    write_mat("mu",    mu_m);
    write_mat("sigma", sig_m);
}

// Compute mean of each row
vec rowwise_mean(const mat& M) {
    vec out(M.rows());
    for (int i = 0; i < M.rows(); ++i)
        out[i] = mean(M.get_row(i));
    return out;
}

// Compute standard deviation of each row
vec rowwise_std(const mat& M, const vec& mu) {
    vec out(M.rows());
    for (int i = 0; i < M.rows(); ++i) {
        vec d = M.get_row(i) - mu[i];
        out[i] = sqrt(sum(sqr(d)) / M.cols());
    }
    return out;
}

int main() {
    mat all_patches;
    // Collect patches from a dataset
    for (auto& entry : fs::directory_iterator(DATASET_DIR)) {
        cv::Mat img = cv::imread(entry.path().string(), cv::IMREAD_GRAYSCALE);
        if (img.rows < PATCH_SIZE || img.cols < PATCH_SIZE) continue;
        mat P = extract_patches(img, PATCH_SIZE, PATCHES_PER_IMG);
        all_patches = all_patches.rows() ? concat_vertical(all_patches, P) : P;
    }
    mat data = all_patches.transpose();  

    // Run FastICA
    Fast_ICA ica(data);
    ica.set_nrof_independent_components(NUM_COMPONENTS);
    ica.set_approach(FICA_APPROACH_SYMM);
    ica.set_non_linearity(FICA_NONLIN_TANH);
    ica.set_max_num_iterations(200);
    ica.set_epsilon(1e-4);
    ica.separate();

    // Compute the independent coefficients for all patches
    mat basis = ica.get_separating_matrix();
    mat comps = ica.get_independent_components();
    vec mu    = rowwise_mean(comps);
    vec sigma = rowwise_std(comps, mu);

    // Save the model
    save_yaml(OUTPUT_FILE, basis, mu, sigma);
    
    return 0;
}
